File: Sources/OSX/NSStoryboard+Swizzling.m
//
//  NSStoryboard+Swizzling.m
//  SwinjectStoryboard
//
//  Created by Mark DiFranco on 2017-03-31.
//  Copyright © 2017 Swinject Contributors. All rights reserved.
//

#import "NSStoryboard+Swizzling.h"
#import <objc/runtime.h>
#import <SwinjectStoryboard/SwinjectStoryboard-Swift.h>
#import <SwinjectStoryboard/SwinjectStoryboardProtocol.h>

@implementation NSStoryboard (Swizzling)

+ (void)load {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        Class class = object_getClass((id)self);

        SEL originalSelector = @selector(storyboardWithName:bundle:);
        SEL swizzledSelector = @selector(swinject_storyboardWithName:bundle:);

        Method originalMethod = class_getClassMethod(class, originalSelector);
        Method swizzledMethod = class_getClassMethod(class, swizzledSelector);

        BOOL didAddMethod =
        class_addMethod(class,
                        originalSelector,
                        method_getImplementation(swizzledMethod),
                        method_getTypeEncoding(swizzledMethod));

        if (didAddMethod) {
            class_replaceMethod(class,
                                swizzledSelector,
                                method_getImplementation(originalMethod),
                                method_getTypeEncoding(originalMethod));
        } else {
            method_exchangeImplementations(originalMethod, swizzledMethod);
        }
    });
}

+ (nonnull instancetype)swinject_storyboardWithName:(NSString *)name bundle:(nullable NSBundle *)storyboardBundleOrNil {
    if (self == [NSStoryboard class]) {
        
        // Instantiate SwinjectStoryboard if NSStoryboard is trying to be instantiated.
        if ([SwinjectStoryboard isCreatingStoryboardReference]) {
            return [SwinjectStoryboard createReferencedWithName:name bundle:storyboardBundleOrNil];
        } else {
            return [SwinjectStoryboard createWithName:name bundle:storyboardBundleOrNil];
        }
    } else {
        return [self swinject_storyboardWithName:name bundle:storyboardBundleOrNil];
    }
}

@end

File: Sources/OSX/NSStoryboard+Swizzling.swift
#if os(OSX)
import Cocoa

extension NSStoryboard {
    static func swizzling() {
        DispatchQueue.once(token: "swinject.storyboard.init") {
            let aClass: AnyClass = object_getClass(self)!

            let originalSelector = #selector(NSStoryboard.init(name:bundle:))
            let swizzledSelector = #selector(swinject_init(name:bundle:))

            let originalMethod = class_getInstanceMethod(aClass, originalSelector)!
            let swizzledMethod = class_getInstanceMethod(aClass, swizzledSelector)!

            let didAddMethod = class_addMethod(aClass, originalSelector,
                                               method_getImplementation(swizzledMethod),
                                               method_getTypeEncoding(swizzledMethod))

            guard didAddMethod else {
                method_exchangeImplementations(originalMethod, swizzledMethod)
                return
            }
            class_replaceMethod(aClass, swizzledSelector,
                                method_getImplementation(originalMethod),
                                method_getTypeEncoding(originalMethod))
        }
    }

    @objc class func swinject_init(name: String, bundle: Bundle?) -> NSStoryboard {
        guard self == NSStoryboard.self else {
            return self.swinject_init(name: name, bundle: bundle)
        }
        // Instantiate SwinjectStoryboard if NSStoryboard is trying to be instantiated.
        if SwinjectStoryboard.isCreatingStoryboardReference {
            return SwinjectStoryboard.createReferenced(name: name, bundle: bundle)
        } else {
            return SwinjectStoryboard.create(name: name, bundle: bundle)
        }
    }
}
#endif
